### PR TITLE
🌱  internal/topology migrate reconcile_state tests to envtest

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

This improves the testing for the topology controller by using a real API Server instead of the fakeClient.

* Separates tests by randomised namespaces.
* Improves output of some tests by moving `g` inside `t.Run`.
* Uses names generated by Topology controller inside the tests.
* Tests depending on `OwnerReferences` also include UID now.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: internal/controllers/topology/cluster/reconcile_state_test.go [hunk 1]
```diff
@@ -48,17 +46,27 @@ var (
...
+	}
 )
 
 func TestReconcileShim(t *testing.T) {
 	infrastructureCluster := builder.InfrastructureCluster(metav1.NamespaceDefault, "infrastructure-cluster1").Build()
 	controlPlane := builder.ControlPlane(metav1.NamespaceDefault, "infrastructure-cluster1").Build()
-	cluster1 := builder.Cluster(metav1.NamespaceDefault, "cluster1").Build()
-	cluster1Shim := clusterShim(cluster1)
+	cluster := builder.Cluster(metav1.NamespaceDefault, "cluster1").Build()
+	cluster.SetUID("foo")
```
Comment (fabriziopandini) [inline internal/controllers/topology/cluster/reconcile_state_test.go:58]: Can we add a comment documenting why we need to explicitly set UID?

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
@killianmuldoon @sbueringer PTAL

File: internal/controllers/topology/cluster/reconcile_state_test.go [hunk 2]
```diff
@@ -48,17 +46,27 @@ var (
...
 
 	t.Run("Shim gets created when InfrastructureCluster and ControlPlane object have to be created", func(t *testing.T) {
 		g := NewWithT(t)
 
+		// Create namespace and modify input to have correct namespace set
+		namespace, err := env.CreateNamespace(ctx, "reconcile-cluster-shim")
+		g.Expect(err).To(Succeed())
```
Comment (sbueringer) [inline internal/controllers/topology/cluster/reconcile_state_test.go:65]: nit / just fyi
In general we're using To/ToNot(HaveOccured) for errors and To/ToNot(Succeed) for func calls (similar below)

**Review by sbueringer — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
Thx!
/lgtm

**Pull request merged**